shell/common/surface.cc

```diff
@@ -3,12 +3,20 @@
 // found in the LICENSE file.
 
 #include "flutter/shell/common/surface.h"
+#include "third_party/skia/include/core/SkSurface.h"
 
 namespace shell {
 
-SurfaceFrame::SurfaceFrame() : submitted_(false) {}
+SurfaceFrame::SurfaceFrame(sk_sp<SkSurface> surface,
+                           SubmitCallback submit_callback)
+    : submitted_(false), surface_(surface), submit_callback_(submit_callback) {}
 
-SurfaceFrame::~SurfaceFrame() = default;
+SurfaceFrame::~SurfaceFrame() {
+  if (submit_callback_) {
+    // Dropping without a Submit.
+    submit_callback_(nullptr);
+  }
+}
 
 bool SurfaceFrame::Submit() {
```

Comment:
Since this is a generic interface to GL, Vulkan and Null backends, it is pretty much impossible for the the caller to make intelligent decisions about a `false` response by looking at just this signature in the header file without any knowledge about the backend in use. Fortunately, this does not happen very often since the users of `SurfaceFrame` are themselves backend specific. I think the assertion in the constructor is a good idea.

In general, `false` just means that the intent of the call did not come to fruition.  Nothing more, nothing less. I will say though that I have an intense dislike for methods that return `void`. Callers at each step in the process of using the API should be able to get immediate feedback on whether that step succeeded. This I think provides more context for error handling. For example, if the `SurfaceFrame` was somehow invalid or unserviceable, `AcquireFrame` itself would refuse to return a valid object (as opposed to returning a `SurfaceFrame` that was somehow misconfigured).